### PR TITLE
Reorganize distributions module

### PR DESCRIPTION
**Pull request by neerajprad**
This is a follow-up to #128 to reorganize the distributions and move the scipy API to contrib. Unfortunately, it shows as a large PR in github but its for the most part just an exchange of files between two modules, i.e. the `continuous`, `discrete` and `distribution` modules in `contrib` are swapped with those in `distributions` module.
 - I had to change the imports and got into a circular dependency issue. To resolve that, I had to merge `constraints`, `constraint_registry` and `transforms` into one module. I think this makes sense since now they don't have any circular dependencies and is mostly self-contained.
 - Changed tests / examples accordingly, and removed the previous hacks in #128 to support both the API. 

I have not made changes to the scipy API in any way, so that should be useable with some minor modifications like I had made in the previous PR.

### CODE REVIEW

**Comment by fehiepsi** [comment]
> it shows as a large PR in github

No worries, it is indeed a good chance to review the whole thing. ;)

**Comment by neerajprad** [comment]
I thought there will be merge conflicts with the other PR; but seems like this is good to review.

File: numpyro/distributions/discrete.py [hunk 1]
```diff
@@ -1,171 +1,393 @@
...
+            self._validate_sample(value)
+        batch_shape = lax.broadcast_shapes(np.shape(value), self.batch_shape)
+        value = np.expand_dims(value, axis=-1)
+        value = np.broadcast_to(value, batch_shape + (1,))
+        log_pmf = np.broadcast_to(self.logits, batch_shape + np.shape(self.logits)[-1:])
+        return np.take_along_axis(log_pmf, value, axis=-1)[..., 0]
+
+    @lazy_property
+    def logits(self):
+        return _to_logits_multinom(self.probs)
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py]: I don't see `logits` property in other discrete distribution. Could you add a comment for why it is needed in the computation of `log_prob`?

File: numpyro/distributions/discrete.py [hunk 2]
```diff
@@ -1,171 +1,393 @@
...
+        super(Poisson, self).__init__(np.shape(rate), validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        return poisson(key, self.rate, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        value = value.astype(get_dtypes(self.rate)[0])
+        return (np.log(self.rate) * value) - self.rate - gammaln(value + 1)
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py]: nit: move `-self.rate` to the end

File: numpyro/distributions/discrete.py [hunk 3]
```diff
@@ -1,171 +1,393 @@
...
+        self.rate = rate
+        super(Poisson, self).__init__(np.shape(rate), validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        return poisson(key, self.rate, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        value = value.astype(get_dtypes(self.rate)[0])
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py]: nit: use `lax.convert_element_dtype` to take care of scalar.

File: numpyro/distributions/discrete.py [hunk 4]
```diff
@@ -1,171 +1,393 @@
...
+    def probs(self):
+        return _to_probs_multinom(self.logits)
+
+    @property
+    def mean(self):
+        return np.broadcast_to(np.nan, self.batch_shape)
+
+    @property
+    def variance(self):
+        return np.broadcast_to(np.nan, self.batch_shape)
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py]: nit: no need for `np.broadcast_to`.

**Comment by neerajprad** [comment]
@fehiepsi - thanks for taking another look at the distributions. I'll address / respond to all your comments shortly.

File: numpyro/distributions/continuous.py [hunk 5]
```diff
@@ -1,441 +1,373 @@
...
+    support = constraints.real
+
+    def __init__(self, loc, scale, validate_args=None):
+        self.loc, self.scale = promote_shapes(loc, scale)
+        batch_shape = lax.broadcast_shapes(np.shape(loc), np.shape(scale))
+        super(Cauchy, self).__init__(batch_shape=batch_shape, validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        u = random.uniform(key, shape=size + self.batch_shape)
+        eps = np.tan(np.pi * (u - 0.5))
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/continuous.py]: Nice!

File: numpyro/distributions/continuous.py [hunk 6]
```diff
@@ -1,441 +1,373 @@
...
+        self.scale = scale
+        super(HalfCauchy, self).__init__(base_dist, AbsTransform(),
+                                         validate_args=validate_args)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        log_prob = self.base_dist.log_prob(value) + np.log(2)
+        value, log_prob = promote_shapes(value, log_prob)
+        log_prob = np.where(value < 0, -np.inf, log_prob)
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/continuous.py]: Hmm..this is an interesting one. I think a reasonable convention and something that we could try to satisfy is that if the values lie outside the support, the corresponding log density evaluation should give a `NaN` or `-inf` or just throw an exception (not nice, but still acceptable). Otherwise the return values could appear to be valid but are actually silent bugs. In this case, since `base_dist` is `cauchy` it will happily return finite log density for negative values which can be problematic (more so since validation check is disabled by default). But you are right that, by that token, `Exponential` should also do this masking. I don't have a good solution for this. Let us remove this, and maybe make validation check enabled by default (will do so separately).

File: numpyro/distributions/continuous.py [hunk 7]
```diff
@@ -1,441 +1,373 @@
...
+
+
+class Beta(Distribution):
+    arg_constraints = {'concentration1': constraints.positive, 'concentration0': constraints.positive}
+    support = constraints.unit_interval
+
+    def __init__(self, concentration1, concentration0, validate_args=None):
+        batch_shape = lax.broadcast_shapes(np.shape(concentration1), np.shape(concentration0))
+        self.concentration1 = np.broadcast_to(concentration1, batch_shape)
+        self.concentration0 = np.broadcast_to(concentration0, batch_shape)
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/continuous.py]: `promote_shapes` isn't sufficient here since we have to stack `conc1`, `conc0` which don't  broadcast internally, so a full broadcasting is unfortunately needed. I think we can have it return the full batch shape too. `lax.broadcast_shapes` should be really cheap though, so I'm not sure if its worth the effort.

File: numpyro/distributions/discrete.py [hunk 8]
```diff
@@ -1,171 +1,393 @@
...
+        return multinomial_rvs(key, self.total_count, self.probs, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        dtype = get_dtypes(self.probs)[0]
+        value = lax.convert_element_type(value, dtype)
+        total_count = lax.convert_element_type(self.total_count, dtype)
+        logits = self.logits
+        logits = np.clip(logits, a_min=np.finfo(get_dtypes(logits)[0]).min)
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: This is needed for numerical stability in `log_prob` otherwise with a value of `0`, we might end up multiplying `0 x -inf` leading to NaNs. This is to ensure that this evaluates to `0`.

File: numpyro/distributions/discrete.py (hunk 4 above)
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: We need the shape to be the same as `batch_shape` but we can use `lax.full` too. Will change.

File: numpyro/distributions/discrete.py (hunk 1 above)
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: `logits` is nothing but a clipped version of `np.log(probs)` for the multinomial case. The other distributions don't have it because they use the `xlogy` formulation in `log_prob`. That said, it might be a good idea to have this in other distributions too, since `log` is expensive and if this can be cached across multiple calls to `log_prob` we should do that.

File: numpyro/distributions/continuous.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/distributions/continuous.py]: Ah, I thought that `promote_shapes` returns broadcasted tensors. So many comments from me are invalid. >"< Sorry about that!

File: numpyro/distributions/continuous.py [hunk 9]
```diff
@@ -1,441 +1,373 @@
...
+    def sample(self, key, size=()):
+        shape = size + self.batch_shape + self.event_shape
+        return standard_gamma(key, self.concentration, shape=shape) / self.rate
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        normalize_term = (self.concentration * np.log(self.rate) -
+                          self.rate * value - gammaln(self.concentration))
+        return (self.concentration - 1) * np.log(value) + normalize_term
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/distributions/continuous.py]: Ah, I miss this point: we should move `self.rate * value` out of `normalize_term` too

File: numpyro/distributions/continuous.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/distributions/continuous.py]: You are right! This is more involved than I thought because negative value can provide a valid log_prob here. Same for Exponential. I think that it is good to keep this implementation as is and do the mask for each distribution which has similar issue. :)

File: numpyro/distributions/discrete.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/distributions/discrete.py]: But `logits` constraint is real (not extended real) so `nan` is a good value to return in that case IMO.

File: numpyro/distributions/continuous.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/continuous.py]: I think the larger issue is that in numpy/jax, we don't have something like `torch.expand` which returns a broadcasted array but doesn't modify the underlying memory (just returns a view). Therefore, we shouldn't naively do full broadcasting unless really needed, like in this case. `promote_shapes` just reshapes the arrays and does not need to copy the memory.

File: numpyro/distributions/discrete.py (hunk 4 above)
Comment (fehiepsi) [reply numpyro/distributions/discrete.py]: >"< I don't know why I make a comment for it here. Srr.

File: numpyro/distributions/discrete.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/distributions/discrete.py]: Hmm, I expect that there are many things to cache but currently we don't use that convention. For example, `gammaln(count)` is also expensive, `log(scale)` in normal/cauchy distributions are also expensive. Could you please remind me why we need to separate `Multinomial` and `MultinomialWithLogits`? (so I can decide what to do with MVN distribution)

File: numpyro/distributions/continuous.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/distributions/continuous.py]: It is good to know! Thanks @neerajprad !

File: numpyro/distributions/continuous.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/distributions/continuous.py]: Sorry, how about going with your decision (to remove this mask) and taking care of this separately? Some distributions such as LogNormal is not easy to justify at a glance.

File: numpyro/distributions/discrete.py (hunk 1 above)
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: I agree with you - let us remove this and if we really need to cache computations, we should do it in a more principled way. 

> Could you please remind me why we need to separate Multinomial and MultinomialWithLogits? (so I can decide what to do with MVN distribution)

We don't need to (and we aren't in PyTorch) - it just seemed easier to not have to sprinkle the code with `if / else`, and have to put in optional parameters in `arg_contraints`. I think, within limits, if we have multiple parametrizations for a distribution, it is best to separate out the implementations. e.g. `OneHotCategorical` and `Categorical`, for instance. How is this relevant to MVN?

File: numpyro/distributions/discrete.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: This was to ensure that the behavior is consistent with what the `xlogy` implementation does effectively, but you are right that this should be the caller's responsibility instead.

File: numpyro/distributions/discrete.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/distributions/discrete.py]: In MVN, there are 3 args `scale_tril`, `covariance`, `precision` so there are 3 constraints to check. If we don't use `precision_matrix` then we should not check for it because it will be expensive taking inverse. I think that we just check for all of them as in PyTorch and ask users for providing suitable `validate_args` argument if they want to filter out some constraints.

**Comment by fehiepsi** [comment]
@neerajprad This is a great PR! But it is big so it is hard to track down discussions. Please merge this when it is ready to you. We can address other nits from me in a later PR.

**Comment by neerajprad** [comment]
> But it is big so it is hard to track down discussions. 

That's the reason why I did it in stages, and this last part was just supposed to be a folder reorganization; but thanks to your review, it looks better than what we had originally. I think I should have addressed most of your nits, but let us follow up separately if you notice other issues.

**Comment by fehiepsi** [comment]
Yup, I just thought that it is convenience to review a whole thing at a time. But the discussions turn out so large to handle in 1 PR. It is so inconvenience for you! >"<

File: numpyro/distributions/discrete.py (hunk 1 above)
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: Hmm..the the only flip-side to having separate classes is code duplication. This is one of those cases where I think some code duplication might be okay (within limits, we don't want 10 MVN classes that end up confusing the user 😄), since it is not like we will be updating the distributions code every day. The advantage is that we don't have to figure about which property was provided and which property should not accessed (via `lazy_property`), etc. 

I will let you figure out how you want to go about it, either is fine with me, and I'm open to merging the logits/probs classes too.

File: numpyro/distributions/continuous.py [hunk 10]
```diff
@@ -1,441 +1,373 @@
...
-pareto = pareto_gen(a=1.0, name="pareto")
-t = t_gen(name='t')
-trunccauchy = trunccauchy_gen(name='trunccauchy')
-truncnorm = truncnorm_gen(name='truncnorm')
-uniform = uniform_gen(a=0.0, b=1.0, name='uniform')
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        return -((value - self.loc) ** 2) / (2.0 * self.scale ** 2) \
+               - np.log(self.scale) - np.log(np.sqrt(2 * np.pi))
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/continuous.py]: Oh..sorry! All of these comments were collapsed and I didn't notice them. 😆 Let me address these, otherwise we'll lose them.

File: numpyro/distributions/continuous.py [hunk 11]
```diff
@@ -1,441 +1,373 @@
...
+        batch_shape = lax.broadcast_shapes(np.shape(scale), np.shape(alpha))
+        self.scale, self.alpha = np.broadcast_to(scale, batch_shape), np.broadcast_to(alpha, batch_shape)
+        base_dist = Exponential(self.alpha)
+        transforms = [ExpTransform(), AffineTransform(loc=0, scale=self.scale)]
+        super(Pareto, self).__init__(base_dist, transforms, validate_args=validate_args)
+
+    @property
+    def mean(self):
+        # mean is inf for alpha <= 1
+        a = lax.div(self.alpha * self.scale, (self.alpha - 1))
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/continuous.py]: native division here would throw an exception when `alpha = 1`, whereas `lax.div` is safe.

File: numpyro/distributions/continuous.py [hunk 12]
```diff
@@ -1,441 +1,373 @@
...
+    def sample(self, key, size=()):
+        size = size + self.batch_shape
+        return self.low + random.uniform(key, shape=size) * (self.high - self.low)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        within_bounds = ((value >= self.low) & (value < self.high))
+        return np.log(lax.convert_element_type(within_bounds, get_dtypes(self.low)[0])) - \
+            np.log(self.high - self.low)
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/continuous.py]: Yeah, lets figure out later if we need to do such masking. Will remove.

File: numpyro/distributions/discrete.py [hunk 13]
```diff
@@ -1,171 +1,393 @@
...
+        self.probs = probs
+        super(Bernoulli, self).__init__(batch_shape=np.shape(self.probs), validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        return random.bernoulli(key, self.probs, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        return xlogy(value, self.probs) + xlog1py(1-value, -self.probs)
```
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: flake8 accepts this within function args, I think. Will change though.

File: numpyro/distributions/continuous.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/distributions/continuous.py]: This does not happen in my system though. Maybe this is fixed in newer version of JAX.

[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/continuous.py]: Hmm..this is a python thing. Note that `np.broadcast_to(1, ()) / np.broadcast_to(0, ())` would give a `ZeroDivisionError` since those values are not jax scalars but python floats.

File: numpyro/distributions/continuous.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (neerajprad) [reply numpyro/distributions/continuous.py]: Oops..that's my bad. Will fix tall of these.

File: numpyro/distributions/continuous.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/distributions/continuous.py]: Nice catch! This is good to learn!
